File: apps/mobile/lib/priority_engine.dart
import 'dart:math' as math;
import 'comprehensive_report_models.dart';

/// Comprehensive result of the explainable civic priority analysis
class CivicPriorityAnalysis {
  final double score; // 0.0 - 100.0
  final ReportPriority mappedPriority;
  final String levelLabel; // 'Critical', 'High', 'Medium', 'Low'
  final List<String> explainableReasons;
  final Map<String, double> signalBreakdown;

  const CivicPriorityAnalysis({
    required this.score,
    required this.mappedPriority,
    required this.levelLabel,
    required this.explainableReasons,
    required this.signalBreakdown,
  });

  /// Formatted score string (e.g. "78/100")
  String get scoreDisplay => '${score.round()}/100';

  /// Top 2-3 explainable summary reasons
  List<String> get topReasons => explainableReasons.take(3).toList();

  Map<String, dynamic> toJson() {
    return {
      'score': double.parse(score.toStringAsFixed(1)),
      'level': levelLabel,
      'mapped_priority': mappedPriority.value,
      'reasons': explainableReasons,
      'breakdown': signalBreakdown,
    };
  }
}

/// Enterprise Civic Smart Prioritization Engine
class CivicPriorityEngine {
  // Decision-support signal weights (Sum = 1.0)
  static const double weightSeverity = 0.30;
  static const double weightPublicSafety = 0.25;
  static const double weightRelatedComplaints = 0.20;
  static const double weightAge = 0.15;
  static const double weightCategory = 0.10;

  // Thresholds for priority level mapping
  static const double thresholdCritical = 80.0;
  static const double thresholdHigh = 60.0;
  static const double thresholdMedium = 35.0;

  /// High-risk keywords indicating immediate public hazard
  static const Set<String> _criticalHazardKeywords = {
    'manhole', 'open manhole', 'wire', 'live wire', 'transformer', 'sparking',
    'electric shock', 'collapse', 'collapsed', 'cave in', 'caved', 'fire',
    'smoke', 'gas leak', 'pipeline burst', 'major flood', 'deep crater',
    'accident', 'emergency', 'fallen tree', 'hanging wire'
  };

  /// Moderate-risk keywords indicating heightened safety concern
  static const Set<String> _moderateHazardKeywords = {
    'pothole', 'broken divider', 'dark street', 'blackout', 'streetlight dead',
    'waterlogging', 'sewage leak', 'garbage burning', 'blocked drain', 'slippery'
  };

  /// Calculate severity score (0 - 100) from AI triage or manual input
  static double calculateSeverityScore(String? severity) {
    if (severity == null || severity.trim().isEmpty) return 50.0; // Neutral baseline

    final s = severity.trim().toLowerCase();
    if (s.contains('critical')) return 100.0;
    if (s.contains('high')) return 80.0;
    if (s.contains('medium')) return 50.0;
    if (s.contains('low')) return 20.0;
    return 50.0;
  }

  /// Calculate public safety impact score (0 - 100) from category, title, and description
  static double calculatePublicSafetyScore({
    String? category,
    String? title,
    String? description,
  }) {
    final combined = '${category ?? ""} ${title ?? ""} ${description ?? ""}'.toLowerCase();

    // Check for critical hazards
    for (final kw in _criticalHazardKeywords) {
      if (combined.contains(kw)) return 100.0;
    }

    // Check for moderate hazards
    for (final kw in _moderateHazardKeywords) {
      if (combined.contains(kw)) return 65.0;
    }

    // Category-based safety baseline
    final cat = (category ?? '').toLowerCase();
    if (cat.contains('safety') || cat.contains('emergency') || cat.contains('disaster')) {
      return 85.0;
    }
    if (cat.contains('electricity') || cat.contains('water') || cat.contains('road')) {
      return 50.0;
    }

    return 25.0; // Standard civic maintenance
  }

  /// Calculate related complaint cluster score (0 - 100)
  static double calculateRelatedComplaintsScore(int relatedCount) {
    if (relatedCount <= 0) return 0.0;
    if (relatedCount == 1) return 40.0;
    if (relatedCount == 2) return 70.0;
    return 100.0; // 3 or more related reports
  }

  /// Calculate complaint age / duration escalation score (0 - 100)
  static double calculateAgeScore(DateTime? createdAt, {bool isResolved = false}) {
    if (isResolved || createdAt == null) return 0.0;

    final diffSeconds = DateTime.now().difference(createdAt).inSeconds;
    if (diffSeconds < 0) return 10.0; // Future/clock skew safe baseline

    final diffDays = diffSeconds / 86400.0;

    if (diffDays < 1.0) return 15.0; // Fresh (< 24 hours)
    if (diffDays <= 3.0) return 40.0; // 1 to 3 days
    if (diffDays <= 7.0) return 70.0; // 4 to 7 days (approaching SLA breach)
    return 100.0; // Over 7 days unresolved (urgent escalation)
  }

  /// Calculate category baseline score (0 - 100)
  static double calculateCategoryBaselineScore(String? category) {
    if (category == null || category.trim().isEmpty) return 50.0;

    final c = category.trim().toLowerCase();
    if (c.contains('safety') || c.contains('emergency') || c.contains('disaster')) {
      return 100.0;
    }
    if (c.contains('electricity') || c.contains('water') || c.contains('drainage')) {
      return 75.0;
    }
    if (c.contains('road') || c.contains('traffic') || c.contains('pavement')) {
      return 60.0;
    }
    if (c.contains('waste') || c.contains('garbage') || c.contains('sanitation')) {
      return 45.0;
    }
    return 35.0;
  }

  /// Evaluate complete multi-factor civic priority for a complaint
  static CivicPriorityAnalysis evaluatePriority({
    String? severity,
    String? category,
    String? title,
    String? description,
    int relatedComplaintsCount = 0,
    DateTime? createdAt,
    bool isResolved = false,
  }) {
    final severityScore = calculateSeverityScore(severity);
    final publicSafetyScore = calculatePublicSafetyScore(
      category: category,
      title: title,
      description: description,
    );
    final relatedScore = calculateRelatedComplaintsScore(relatedComplaintsCount);
    final ageScore = calculateAgeScore(createdAt, isResolved: isResolved);
    final categoryScore = calculateCategoryBaselineScore(category);

    final totalScore = (severityScore * weightSeverity) +
        (publicSafetyScore * weightPublicSafety) +
        (relatedScore * weightRelatedComplaints) +
        (ageScore * weightAge) +
        (categoryScore * weightCategory);

    final boundedScore = math.max(0.0, math.min(100.0, totalScore));

    // Map to levels and existing ReportPriority enum
    String levelLabel;
    ReportPriority mappedPriority;

    if (boundedScore >= thresholdCritical) {
      levelLabel = 'Critical';
      mappedPriority = ReportPriority.high; // Maps safely to existing high enum in DB
    } else if (boundedScore >= thresholdHigh) {
      levelLabel = 'High';
      mappedPriority = ReportPriority.high;
    } else if (boundedScore >= thresholdMedium) {
      levelLabel = 'Medium';
      mappedPriority = ReportPriority.medium;
    } else {
      levelLabel = 'Low';
      mappedPriority = ReportPriority.low;
    }

    final reasons = <String>[];

    if (severityScore >= 80.0) {
      reasons.add('${severity ?? "High"} severity identified by AI triage');
    }

    if (publicSafetyScore >= 85.0) {
      reasons.add('High public-safety impact / critical municipal hazard');
    } else if (publicSafetyScore >= 60.0) {
      reasons.add('Elevated neighborhood safety concern');
    }

    if (relatedComplaintsCount >= 3) {
      reasons.add('$relatedComplaintsCount related complaints clustered nearby');
    } else if (relatedComplaintsCount >= 1) {
      reasons.add('$relatedComplaintsCount related complaint reported in area');
    }

    if (ageScore >= 70.0 && !isResolved) {
      final days = createdAt != null ? (DateTime.now().difference(createdAt).inDays) : 4;
      reasons.add('Unresolved for ${math.max(days, 4)} days (SLA escalation)');
    }

    if (categoryScore >= 75.0) {
      reasons.add('Critical infrastructure category ($category)');
    }

    if (reasons.isEmpty) {
      reasons.add('Standard priority assessment for general civic grievance');
    }

    return CivicPriorityAnalysis(
      score: boundedScore,
      mappedPriority: mappedPriority,
      levelLabel: levelLabel,
      explainableReasons: reasons,
      signalBreakdown: {
        'severity': severityScore,
        'public_safety': publicSafetyScore,
        'related_complaints': relatedScore,
        'age_escalation': ageScore,
        'category_baseline': categoryScore,
      },
    );
  }

  /// Convenience helper to calculate priority directly from a ComprehensiveReportModel
  static CivicPriorityAnalysis evaluateReport(ComprehensiveReportModel report, {int relatedCount = 0}) {
    final isResolved = report.status == ReportStatus.resolved;
    final effectiveRelated = relatedCount > 0 ? relatedCount : report.consolidatedReports;

    return evaluatePriority(
      severity: report.priority.displayName,
      category: report.category,
      title: report.title,
      description: report.description,
      relatedComplaintsCount: effectiveRelated,
      createdAt: report.createdAt,
      isResolved: isResolved,
    );
  }
}

File: apps/web2/src/services/priorityEngine.ts
import { Complaint, ComplaintPriority } from '../types/complaint';
import { SimilarityEngine } from './similarityEngine';

export type PriorityLevel = 'Critical' | 'High' | 'Medium' | 'Low';

export interface PrioritySignalBreakdown {
  severity: number;
  publicSafety: number;
  relatedComplaints: number;
  ageEscalation: number;
  categoryBaseline: number;
}

export interface CivicPriorityAnalysis {
  score: number; // 0.0 - 100.0
  levelLabel: PriorityLevel;
  scoreDisplay: string; // e.g. "82/100"
  badgeColor: string;
  badgeBg: string;
  badgeBorder: string;
  badgeText: string;
  explainableReasons: string[];
  topDrivers: string[];
  signalBreakdown: PrioritySignalBreakdown;
}

export class PriorityEngine {
  // Explainable signal weights (Sum = 1.0)
  public static readonly weightSeverity = 0.30;
  public static readonly weightPublicSafety = 0.25;
  public static readonly weightRelatedComplaints = 0.20;
  public static readonly weightAge = 0.15;
  public static readonly weightCategory = 0.10;

  // Thresholds for priority level mapping
  public static readonly thresholdCritical = 80.0;
  public static readonly thresholdHigh = 60.0;
  public static readonly thresholdMedium = 35.0;

  // High-risk keywords indicating immediate public hazard
  private static readonly criticalHazardKeywords = [
    'manhole', 'open manhole', 'wire', 'live wire', 'transformer', 'sparking',
    'electric shock', 'collapse', 'collapsed', 'cave in', 'caved', 'fire',
    'smoke', 'gas leak', 'pipeline burst', 'major flood', 'deep crater',
    'accident', 'emergency', 'fallen tree', 'hanging wire'
  ];

  // Moderate-risk keywords indicating heightened safety concern
  private static readonly moderateHazardKeywords = [
    'pothole', 'broken divider', 'dark street', 'blackout', 'streetlight dead',
    'waterlogging', 'sewage leak', 'garbage burning', 'blocked drain', 'slippery'
  ];

  /**
   * Calculates severity score (0 - 100) from priority string or AI triage
   */
  public static calculateSeverityScore(severity?: string | null): number {
    if (!severity || !severity.trim()) return 50.0; // Neutral baseline

    const s = severity.trim().toLowerCase();
    if (s.includes('critical') || s.includes('urgent')) return 100.0;
    if (s.includes('high')) return 80.0;
    if (s.includes('medium')) return 50.0;
    if (s.includes('low')) return 20.0;
    return 50.0;
  }

  /**
   * Calculates public safety impact score (0 - 100) from category, title, and description
   */
  public static calculatePublicSafetyScore(
    category?: string | null,
    title?: string | null,
    description?: string | null
  ): number {
    const combined = `${category || ''} ${title || ''} ${description || ''}`.toLowerCase();

    // Check for critical hazards
    for (const kw of this.criticalHazardKeywords) {
      if (combined.includes(kw)) return 100.0;
    }

    // Check for moderate hazards
    for (const kw of this.moderateHazardKeywords) {
      if (combined.includes(kw)) return 65.0;
    }

    // Category-based safety baseline
    const cat = (category || '').toLowerCase();
    if (cat.includes('safety') || cat.includes('emergency') || cat.includes('disaster')) {
      return 85.0;
    }
    if (cat.includes('electricity') || cat.includes('water') || cat.includes('road')) {
      return 50.0;
    }

    return 25.0; // Standard municipal maintenance
  }

  /**
   * Calculates related complaint cluster score (0 - 100)
   */
  public static calculateRelatedComplaintsScore(relatedCount: number): number {
    if (relatedCount <= 0) return 0.0;
    if (relatedCount === 1) return 40.0;
    if (relatedCount === 2) return 70.0;
    return 100.0; // 3 or more related reports
  }

  /**
   * Calculates complaint age / duration escalation score (0 - 100)
   */
  public static calculateAgeScore(
    createdAt?: Date | string | null,
    isResolved: boolean = false
  ): number {
    if (isResolved || !createdAt) return 0.0;

    const d = typeof createdAt === 'string' ? new Date(createdAt) : createdAt;
    if (isNaN(d.getTime())) return 10.0;

    const diffSeconds = (Date.now() - d.getTime()) / 1000;
    if (diffSeconds < 0) return 10.0; // Future/clock skew safe baseline

    const diffDays = diffSeconds / 86400.0;

    if (diffDays < 1.0) return 15.0; // Fresh (< 24 hours)
    if (diffDays <= 3.0) return 40.0; // 1 to 3 days
    if (diffDays <= 7.0) return 70.0; // 4 to 7 days (approaching SLA breach)
    return 100.0; // Over 7 days unresolved (urgent escalation)
  }

  /**
   * Calculates category baseline score (0 - 100)
   */
  public static calculateCategoryBaselineScore(category?: string | null): number {
    if (!category || !category.trim()) return 50.0;

    const c = category.trim().toLowerCase();
    if (c.includes('safety') || c.includes('emergency') || c.includes('disaster')) {
      return 100.0;
    }
    if (c.includes('electricity') || c.includes('water') || c.includes('drainage')) {
      return 75.0;
    }
    if (c.includes('road') || c.includes('traffic') || c.includes('pavement')) {
      return 60.0;
    }
    if (c.includes('waste') || c.includes('garbage') || c.includes('sanitation')) {
      return 45.0;
    }
    return 35.0;
  }

  /**
   * Evaluates complete multi-signal civic priority for a Complaint object
   */
  public static evaluateComplaintPriority(
    complaint: Complaint,
    allComplaints?: Complaint[]
  ): CivicPriorityAnalysis {
    const isResolved =
      complaint.status === 'resolved' ||
      complaint.status === 'verified' ||
      complaint.status === 'closed';

    // Phase 3A integration: calculate real related complaints count
    let relatedCount = 0;
    if (allComplaints && allComplaints.length > 0) {
      const relatedList = SimilarityEngine.findRelatedComplaints(complaint, allComplaints, {
        limit: 10,
        minConfidence: SimilarityEngine.thresholdRelatedIncident,
      });
      relatedCount = relatedList.length;
    }

    const severityScore = this.calculateSeverityScore(complaint.priority);
    const publicSafetyScore = this.calculatePublicSafetyScore(
      complaint.rawCategory || complaint.category,
      complaint.title,
      complaint.description
    );
    const relatedScore = this.calculateRelatedComplaintsScore(relatedCount);
    const ageScore = this.calculateAgeScore(complaint.createdAt, isResolved);
    const categoryScore = this.calculateCategoryBaselineScore(
      complaint.rawCategory || complaint.category
    );

    const totalScore =
      severityScore * this.weightSeverity +
      publicSafetyScore * this.weightPublicSafety +
      relatedScore * this.weightRelatedComplaints +
      ageScore * this.weightAge +
      categoryScore * this.weightCategory;

    const boundedScore = Math.max(0.0, Math.min(100.0, totalScore));

    let levelLabel: PriorityLevel;
    let badgeColor: string;
    let badgeBg: string;
    let badgeBorder: string;
    let badgeText: string;

    if (boundedScore >= this.thresholdCritical) {
      levelLabel = 'Critical';
      badgeColor = '#D92D20';
      badgeBg = 'bg-red-50';
      badgeBorder = 'border-red-200';
      badgeText = 'text-red-700';
    } else if (boundedScore >= this.thresholdHigh) {
      levelLabel = 'High';
      badgeColor = '#EA580C';
      badgeBg = 'bg-orange-50';
      badgeBorder = 'border-orange-200';
      badgeText = 'text-orange-700';
    } else if (boundedScore >= this.thresholdMedium) {
      levelLabel = 'Medium';
      badgeColor = '#D99A00';
      badgeBg = 'bg-amber-50';
      badgeBorder = 'border-amber-200';
      badgeText = 'text-amber-700';
    } else {
      levelLabel = 'Low';
      badgeColor = '#1769D2';
      badgeBg = 'bg-blue-50';
      badgeBorder = 'border-blue-200';
      badgeText = 'text-blue-700';
    }

    const reasons: string[] = [];

    if (publicSafetyScore >= 85.0) {
      reasons.push('Immediate public safety risk detected in issue description');
    } else if (publicSafetyScore >= 60.0) {
      reasons.push('Elevated neighborhood safety concern');
    }

    if (severityScore >= 80.0) {
      reasons.push(`${complaint.priority.toUpperCase()} priority tier`);
    }

    if (relatedCount >= 2) {
      reasons.push(`${relatedCount} related citizen grievances clustered in vicinity`);
    } else if (relatedCount === 1) {
      reasons.push('1 related complaint reported in area');
    }

    if (ageScore >= 70.0 && !isResolved) {
      const createdDate = new Date(complaint.createdAt);
      const days = !isNaN(createdDate.getTime())
        ? Math.max(1, Math.round((Date.now() - createdDate.getTime()) / 86400000))
        : 4;
      reasons.push(`Long pending (${days} days unresolved · SLA escalation)`);
    }

    if (categoryScore >= 75.0) {
      reasons.push(`Critical infrastructure domain (${complaint.categoryLabel || complaint.category})`);
    }

    if (reasons.length === 0) {
      reasons.push('Standard municipal maintenance grievance');
    }

    const topDrivers: string[] = [];
    if (publicSafetyScore >= 65.0) topDrivers.push('Safety Risk');
    if (relatedCount >= 2) topDrivers.push(`${relatedCount} Clustered Reports`);
    if (ageScore >= 70.0 && !isResolved) topDrivers.push('Long Pending');
    if (severityScore >= 80.0 && topDrivers.length < 3) topDrivers.push('High Severity');
    if (topDrivers.length === 0) topDrivers.push('Routine Maintenance');

    return {
      score: boundedScore,
      levelLabel,
      scoreDisplay: `${Math.round(boundedScore)}/100`,
      badgeColor,
      badgeBg,
      badgeBorder,
      badgeText,
      explainableReasons: reasons,
      topDrivers,
      signalBreakdown: {
        severity: severityScore,
        publicSafety: publicSafetyScore,
        relatedComplaints: relatedScore,
        ageEscalation: ageScore,
        categoryBaseline: categoryScore,
      },
    };
  }

  /**
   * Sorts a collection of complaints deterministically by calculated Civic Priority score (descending),
   * using age / creation date as secondary tie-breaker.
   */
  public static sortComplaintsByPriority(complaints: Complaint[]): Array<{
    complaint: Complaint;
    priorityAnalysis: CivicPriorityAnalysis;
  }> {
    const analyzed = complaints.map((c) => ({
      complaint: c,
      priorityAnalysis: this.evaluateComplaintPriority(c, complaints),
    }));

    return analyzed.sort((a, b) => {
      if (b.priorityAnalysis.score !== a.priorityAnalysis.score) {
        return b.priorityAnalysis.score - a.priorityAnalysis.score;
      }
      // Tie-breaker: older unresolved reports come first
      return (
        new Date(a.complaint.createdAt).getTime() -
        new Date(b.complaint.createdAt).getTime()
      );
    });
  }
}
